### PR TITLE
Clean up return->call address translation for backtraces

### LINKED ISSUE #29695: problems with interpreter backtraces
Seems to be two, possibly disjoint problems observed here:
1. "repeated 387 more times" and the immediately subsequent "repeated 166 more times" are precisely the same stacktrace
2. the interpreter duplicates `f28442` on the callstack

```julia
$ ./julia --compile=min -q
julia> begin
           f28442(c) = g28442(c + 1)
           g28442(c) = c > 2000 ? (return backtrace()) : f28442(c + 1)
           bt = f28442(1)
           Base.show_backtrace(stdout, bt)
       end

Stacktrace:
 [1] g28442(::Int64) at REPL[0]:3
 [2] f28442(::Int64) at REPL[0]:2
 ... (the last 2 lines are repeated 442 more times)
 [887] g28442(::Int64) at REPL[0]:3
 [888] f28442(::Int64) at REPL[0]:2 (repeats 2 times)
 [889] g28442(::Int64) at REPL[0]:3
 ... (the last 2 lines are repeated 387 more times)
 [1664] f28442(::Int64) at REPL[0]:2 (repeats 2 times)
 [1665] g28442(::Int64) at REPL[0]:3
 ... (the last 2 lines are repeated 166 more times)
 [1998] f28442(::Int64) at REPL[0]:2
 [1999] top-level scope at REPL[0]:4
 [2000] eval(::Module, ::Any) at ./boot.jl:319
 [2001] eval_user_input(::Any, ::REPL.REPLBackend) at REPL.jl:85
 [2002] macro expansion at REPL.jl:117 [inlined]
```

### PR DESCRIPTION
**Pull request by c42f**
Here I translate return->call address internally in `jl_unw_stepn` so that users never have to deal with this subtlety when calling `StackTraces.lookup`, as discussed on slack with @vtjnash. This is a prelude to bringing in some larger changes from #33277, but which got conflicted with the merge of #33190.

The first commit also rearranges the `jl_unw_stepn` API a little so that interpreter stack traces can always be fit in to `bt_data` when calling `jl_unw_stepn` incrementally for deep stacks + add test cases for that (fixes #29695). For the tests, clean up printing of `CodeInfo` objects wrapped in `InterpreterIP` as well.

The second commit reverts the API breaking name change lookupat->lookup from #33190. @timholy I think this resolves your reservations from https://github.com/JuliaLang/julia/pull/33190#discussion_r322007723. With return->call address translation done inside `jl_unw_stepn`, there should largely be no need to make the API breaking change, as all backtrace addresses which the user has access to will already be translated and can be passed directly into `StackTraces.lookup()`, exactly as was done in julia 1.3. (There's one possible exception to this rule: If the user were looking up the debug info for a pointer to a C function *not* obtained from a backtrace, they previously would have had to use `StackTraces.lookup(ptr+1)` to undo the the `ptr-1` translation that `StackTraces.lookup` did prior to #33190. This kind of usage is reflected in the test which does `StackTraces.lookup(ctestptr)` but it seems like an edge case and not worth breaking the API over.)

### CODE REVIEW

File: src/stackwalk.c [hunk 1]
```diff
@@ -44,21 +60,58 @@ size_t jl_unw_stepn(bt_cursor_t *cursor, uintptr_t *ip, uintptr_t *sp, size_t ma
...
+            // (For example signal frames on linux x86_64 have `call_ip ==
+            // return_ip`.)
+            //
+            // However for our current purposes it seems sufficient to assume
+            // that `call_ip = return_ip-1`. See also:
+            //
+            // * The LLVM unwinder functions step() and setInfoBasedOnIPRegister()
+            //   https://github.com/llvm/llvm-project/blob/master/libunwind/src/UnwindCursor.hpp
+            // * The way that libunwind handles it in `unw_get_proc_name`:
+            //   https://lists.nongnu.org/archive/html/libunwind-devel/2014-06/msg00025.html
```
Comment (c42f) [inline src/stackwalk.c]: As you can see, I spent a while looking up what people do for this stuff. AFAICT it seems that `return_ip - 1` should be sufficient for our use cases. libunwind does something slightly more fancy, but I guess we don't want to trace through signal handler frames. Also libunwind's `dwarf.use_prev_instr` field is hidden from us behind an opaque pointer in any case (see `unw_get_proc_name`).

The LLVM unwinder seems to have a special case for ARM in thumb mode and I tried to do the same for our ARMv7 support but I don't know how to test that yet. @yuyichao does this use of `_CPU_ARM_` look right?

Comment (yuyichao) [reply src/stackwalk.c]: > As you can see, I spent a while looking up what people do for this stuff. AFAICT it seems that return_ip - 1 should be sufficient for our use cases. libunwind does something slightly more fancy, but I guess we don't want to trace through signal handler frames. Also libunwind's dwarf.use_prev_instr field is hidden from us behind an opaque pointer in any case (see unw_get_proc_name).

As long as we don't give that correct pointer back to libunwind and don't add the offset when we get the backtrace from the signal frame it should be fine.

> does the above use of _CPU_ARM_ look right?

LGTM.

Comment (c42f) [reply src/stackwalk.c]: > don't add the offset when we get the backtrace from the signal frame

Oops, looks like I was doing that part wrong. This should now be cleaned up in the latest commit by avoiding doing the offset for contexts we know have come from a signal handler. I checked that this matches the behavior of libunwind's `unw_backtrace()` for those contexts on my machine.

**Comment by c42f** [comment]
I've added another commit here to be more careful to not adjusting the pointer in the case of signal frames (@vtjnash I think this will also be more correct for use in your Profiler work)

I also generalized the skipping of certain internal frames involved in capturing the backtraces themselves so frames for `jl_backtrace_from_here` and `jl_throw` etc can be systematically elided. (For internal debugging it's possible to put them back in by using a negative `skip` argument to `jl_backtrace_from_here`.)

File: base/stacktraces.jl [hunk 2]
```diff
@@ -96,13 +96,13 @@ end
 
 
 """
-    lookupat(pointer::Union{Ptr{Cvoid}, UInt}) -> Vector{StackFrame}
```
Comment (ararslan) [inline base/stacktraces.jl]: I think we'll need to keep this around with a soft deprecation then hard deprecate it in 2.0.

Comment (KristofferC) [reply base/stacktraces.jl]: This is reverting it back to the original name. There hasn't been a release with this.

Comment (ararslan) [reply base/stacktraces.jl]: Ah, okay, great.

Comment (c42f) [reply base/stacktraces.jl]: Yes; for more context see https://github.com/JuliaLang/julia/pull/33190#discussion_r322007723 and https://github.com/JuliaLang/julia/pull/33190#discussion_r322299303

**Comment by c42f** [comment]
Bump. Any takers for a more detailed look? @JeffBezanson?

Perhaps I should just merge this as it allows us to revert a breaking change, passes tests, and also fixes a long standing bug. But having someone read it closely enough to click "Approve" is always appreciated!

**Review by vchuravy — APPROVED** [review]
looks good to me.

**Comment by c42f** [comment]
Thanks guys!

**Pull request merged**